src/linux.rs

```diff
@@ -30,6 +35,54 @@ pub fn get_mac(name: Option<&str>) -> Result<Option<[u8; 6]>, MacAddressError> {
     Ok(None)
 }
 
+/// Uses the `getifaddrs` call to retrieve a list of network interfaces on the
+/// host device and returns the MAC address that matching the adapter with
+/// the given IP
+///
+/// Because nix returns all of the IP's and MAC's in a combined list, we need
+/// to map the IP to an inteface name and track the MAC addresses by interface name
+/// and see if there's a match
+pub fn get_mac_address_by_ip(ip: &IpAddr) -> Result<Option<[u8; 6]>, MacAddressError> {
+    let ifiter = getifaddrs()?;
+
+    let mut ip_on_inferface: Option<String> = None;
+    let mut mac_to_interface: HashMap<String, Option<[u8; 6]>> = HashMap::new();
+    for interface in ifiter {
+        if let Some(iface_address) = interface.address {
+            // is this a mac address?
+            if let Some(link) = iface_address.as_link_addr() {
+                mac_to_interface.insert(interface.interface_name.clone(), link.addr());
+                // did we just find what we're looking for?
+                if let Some(intf_name) = &ip_on_inferface {
+                    if *intf_name == interface.interface_name {
+                        return Ok(link.addr());
+                    }
+                }
+            }
+            if let Some(adapter_ip) = if let Some(sin4) = iface_address.as_sockaddr_in() {
+                // v4 addr?
+                Some(IpAddr::from(Ipv4Addr::from(sin4.ip())))
+            } else if let Some(sin6) = iface_address.as_sockaddr_in6() {
+                // v6 addr?
+                Some(IpAddr::from(Ipv6Addr::from(sin6.ip())))
+            } else {
+                // something else, ignore
+                None
+            } {
+                // found an IP for this adapter - if it's the one we're looking for, save it
+                if adapter_ip == *ip {
+                    ip_on_inferface = Some(interface.interface_name.clone());
+                    if let Some(mac) = mac_to_interface.get(&interface.interface_name) {
+                        return Ok(mac.clone());
+                    }
+                }
+            }
```

Comment:
This section is pretty confusing, using `if let ...`s for the condition of *another* `if let` directly is also pretty unidiomatic, so I think it'd be better if you bound the IP to a variable first, and `continue` on `None` instead.

Reply:
I disagree about it being un-idiomatic - this is fairly standard functional programming which Rust explicitly tries to support (https://doc.rust-lang.org/book/ch13-00-functional-features.html).  That said, it's your code and I'd rather you agree to land it than argue about it so I can change it.  Let me know if the my response to  your first comment makes sense and then I can tackle any changes together with this.

Reply:
I wouldn't call that style functional, it's an `if ... else if ... else` that evaluates to a `bool`, but regardless, my concern is it took me about ~5 minutes to figure out how that was compiling until I finally broke it down (and I work with Rust almost every day and have seen it before :smile:), its just pretty hard to read IMO with the condition being at the same level as the actual following block, and `} {` being pretty uncommon, and would better be served by a `.map` or `.and_then` on another binding (or even as the conditional).